### DISCUSSION TITLE
Fetching host/port from a Jedis instance (4.X)

### DISCUSSION THREAD

**Discussion by MSPigl** [xetorthio/jedis#3058] [General]
Hello,

I'm currently upgrading my team's codebase from Jedis 2.X to 4.X. In our existing code, we use `Jedis.getClient().getPort()/getHost()`, which is used for logging/unit testing to verify that external configuration matches what the application's internal configuration expects. 

However, 4.X switched `Jedis.getClient()` to just passthrough to `getConnection()`, so I'm not seeing how I would get the host and port for a Jedis instance after the upgrade. The logic that leverages `getHost()/getPort() `isn't application critical and could be reworked if there really is no way to get that data anymore, however I'd definitely prefer that the logic be kept.

Any thoughts on how to retrieve that data in 4.X?

**Comment by sazzad16** [comment]
@MSPigl Thank you for sharing your use case. Let's see if anyone else has anything to say and/or ideas to share.

The inspiration behind the removal of those methods is that the `Connection` may not be bound to traditional host and port anymore. For example we can mention [UDS socket](https://github.com/redis/jedis/blob/69c751d8627e61b9792c30798addb09d792cd579/src/test/java/redis/clients/jedis/UdsTest.java) support. Also, host and port provided in constructor may not be the _true_ ones at all due to the possible presence of `HostAndPortMapper`.

**Comment by YNedderhoff** [comment]
Essentially we have some batching logic to reduce writes. We use a `ShardedJedis` instance to get the shard for a given key, so that we know where it belongs. We batch by using a map with shards as key, and a `LinkedBlockingQueue` of increment operations as value. We define a shard by host and port as a unique identifier. 

Essentially an abstraction layer where for example incrementing will either just add it to the queue, or also flush the queue. A queue exists by shard.

```java
public void incrementBy(IncrementOperation incrementOperation) {
    String shard;
    try (ShardedJedis jedis = shardInfoPool.getResource()) {
        Client client = jedis.getShard(getKey(incrementOperation)).getClient();
        shard = client.getHost() + ":" + client.getPort();
    }

    batchedUpdates.putIfAbsent(shard, new BatchedItem(DateTime.now(), new LinkedBlockingQueue<>()));

    BlockingQueue<IncrementOperation> queue = batchedUpdates.get(shard).getQueue();
    queue.add(incrementOperation);

    maybeFlushQueue(shard, queue);
}
```

So if we could replace host and port with another unique identifier of the shard, I think we'd be fine.

**Reply by sazzad16** [reply]
@YNedderhoff Just to understand correctly, are you describing how you are using Jedis 3.x and asking for a leeway in Jedis 4.x?

**Reply by YNedderhoff** [reply]
Yes. Sorry, could have been clearer.

**Comment by sazzad16** [comment]
@MSPigl @YNedderhoff PTAL at #3137 and let me whether it fulfills your purposes. You'd be able to do `Jedis.getConnection().getSocketString()`.

**Reply by YNedderhoff** [reply]
I haven't tried, but just from the looks of it, I am not sure. Maybe I am wrong though. As you see in the code snippet I shared above, from a `ShardedJedis` instance, I get the correct client for the shard I am looking for using the key. From that client, I then get host and port:

```java
Client client = jedis.getShard(getKey(incrementOperation)).getClient();
shard = client.getHost() + ":" + client.getPort();
```

So there is no `Jedis` instance involved here.

I should maybe clarify that we don't use cluster mode, if that makes a difference.

**Reply by sazzad16** [reply]
`ShardedJedis.getShard(String/byte[] key)` returns a `Jedis` object.

**Reply by YNedderhoff** [reply]
Ah right, that must be another change then in Jedis 4.x.x. It used to return `Client`. Then I believe this change will work well for us 👍  Thank you a lot!

**Reply by sazzad16** [reply]
Wait, NO! That's Jedis 3.x feature. There's no ShardedJedis in Jedis 4.x.

**Reply by sazzad16** [reply]
In Jedis 3.x,
- `ShardedJedis.getShard(String/byte[] key)` returns a `Jedis` object.
- `ShardedJedis.getShard(String/byte[] key).getClient()` returns a `Client` object.

**Reply by YNedderhoff** [reply]
Sorry, yeah, I am using 3.x and I am getting confused. It's been a while since I last tried to upgrade to 4.x.  I would have to look into it then I guess, I can't say from the top of my mind.

**Reply by YNedderhoff** [reply]
Hey @sazzad16 I just looked at our code again, specifically what I shared above:

```java
// Jedis 3.x
public void incrementBy(IncrementOperation incrementOperation) {
    String shard;
    try (ShardedJedis jedis = shardInfoPool.getResource()) {
        Client client = jedis.getShard(getKey(incrementOperation)).getClient();
        shard = client.getHost() + ":" + client.getPort();
    }

    batchedUpdates.putIfAbsent(shard, new BatchedItem(DateTime.now(), new LinkedBlockingQueue<>()));

    BlockingQueue<IncrementOperation> queue = batchedUpdates.get(shard).getQueue();
    queue.add(incrementOperation);

    maybeFlushQueue(shard, queue);
}
```

That `shardInfoPool` there is a `ShardedJedisPool` instance. A little later, we use that very same instance for some pipelining by doing

```java
// Jedis 3.x
try (ShardedJedis resource = shardInfoPool.getResource()) {
    ShardedJedisPipeline pipeline = resource.pipelined();
...
```

Am I right that this now means that I should replace `ShardedJedisPool` with `ShardedPipeline` in Jedis 4.x? Because if so, I I am not sure how the above code would look like with Jedis 4.x and specifically your change proposed in https://github.com/redis/jedis/pull/3137. I want to get the host/shard for a specific key, and for that shard I want to find some unique identifier (like the socket as you suggested). But I don't know how I get the shard, as there's no `getShard` method for `ShardedPipeline`.

**Reply by sazzad16** [reply]
> Am I right that this now means that I should replace `ShardedJedisPool` with `ShardedPipeline` in Jedis 4.x?

@YNedderhoff It seems you should be replacing almost everything with `ShardedPipeline`.

```
ShardedPipeline shardedPipeline;

public void incrementBy(IncrementOperation incrementOperation) {
    insertInPipeline(shardedPipeline, incrementOperation);
}

// shardedPipeline.sync(); // close() calls sync() anyway
shardedPipeline.close();
```

**Reply by YNedderhoff** [reply]
Yes, I thought the same, but, how do I solve my use case as outlined above? The one you're proposing that PR for?

**Reply by sazzad16** [reply]
You won't need that anymore. It'll be handled internally in ShardedPipeline.

**Reply by YNedderhoff** [reply]
I am not following I think. The batched updates I have posted above are handled internally?

**Reply by sazzad16** [reply]
Yes. Each `close()` would complete one batched updates. We may improve it to one batch per `sync()` (#3207) but this may take time.

IMO, ATM a better approach for you could be having one `ShardedConnectionProvider` and use it in multiple pipeline (if possible). This way connections will be better reused.

```
ShardedConnectionProvider provider;
//Pattern pattern; // if you want

// start batch -->
ShardedPipeline pipeline = new ShardedPipeline(provider); // or, new ShardedPipeline(provider, pattern);

//
// add commands/operations in pipeline
//

pipeline.close(); // <-- complete batch
```

**Comment by sazzad16** [comment]
#3745 is just merged which could be useful in OP's use case.
Note: This is targeted for Jedis version 5.2.0.
